overlays/wrap_memory.py

```diff
@@ -0,0 +1,252 @@
+import ast
+import dataclasses
+from typing import (
+    Any, Callable, Dict, Generic, Literal, Protocol, Set, Tuple, TypeVar, List, Optional, cast
+)
+from abc import abstractmethod
+
+from typing_extensions import TypeAlias
+import textwrap
+
+
+T = TypeVar("T")
+
+
+
+class ReadOnlyEnv(Generic[T], Protocol):
+    def __call__(self, key: str, dependency: str) -> T:
+        ...
+
+@dataclasses.dataclass
+class WritableEnv(Generic[T]):
+    # This is only initialized globally. So, it basically stands in for Pyre's shared memory table.
+
+
+    # It's a pain to type this well so I'll place fast and loose
+    # with the types here to avoid an explosion of generics
+    saved_contents_cache_table: Dict[str, T] = dataclasses.field(default_factory=dict)
+    unsaved_contents_cache_table: Dict[str, T] = dataclasses.field(default_factory=dict)
+    unsaved_modules: Set[str] = dataclasses.field(default_factory=set)
+    dependencies: Dict[str, Set[object]] = dataclasses.field(default_factory=dict)
+
+
+def module(key: str) -> str:
+    return key.split(".")[0]
+
+@dataclasses.dataclass
+class EnvTable(Generic[T]):
+    writable_env: WritableEnv[T]
+    upstream_env: Optional["EnvTable"] = None
+
+    def upstream_get(self, use_saved_contents_of_dependents: bool) -> ReadOnlyEnv:
+        if self.upstream_env is None:
+            return ...  # typing this correctly is annoying and not illuminating
+        else:
+            return self.upstream_env.read_only(use_saved_contents_of_dependents)
+
+    @staticmethod
+    def produce_value(key: str, upstream_get: Any, current_env_getter: Any) -> T:
+        "Must be implemented by child environments"
+        raise NotImplementedError()
+
+    def register_dependency(self, key: str, dependency: str) -> None:
+        self.writable_env.dependencies[key] = self.writable_env.dependencies.get(key, set())
+        self.writable_env.dependencies[key].add(dependency)
+
+    def get(self, key: str, dependency: str, use_saved_contents_of_dependents: bool) -> T:
+        self.register_dependency(key, dependency)
+
+        target_cache_table = (self.writable_env.saved_contents_cache_table
+                              if use_saved_contents_of_dependents or module(key) not in self.writable_env.unsaved_modules
+                              else self.writable_env.unsaved_contents_cache_table)
+        # Update the saved_contents_cache_table whether the module is
+        # saved or unsaved.
+        if key not in target_cache_table:
+            target_cache_table[key] = self.produce_value(
+                key,
+                self.upstream_get(use_saved_contents_of_dependents),
```

Comment:
> Having automatically-safe dataflow by completely isolating parent environments from overlays could be very beneficial, making the isolation a manually-coded thing sounds very hard. I might try to draw up a version where we wrap the full environment.

Yeah, that is the approach I started with in earlier commits. I split it up to convince myself that it could work without an explicit fall-back logic and to see the clear structure of the saved contents DAG and the unsaved contents DAG. Need to wrap it up again for software engineering hygiene, but I ran out of time yesterday :P